File: DoublyLinkedList.cpp
#include<bits/stdc++.h>
using namespace std;

class DoublyNode{
public:
    int value;
    DoublyNode* next;
    DoublyNode* prev;

    DoublyNode(int value,DoublyNode* prev = nullptr, DoublyNode* next= nullptr){
        this->value=value;
        this->next = next;
        this->prev=prev;
    }

};

class DoublyLinkedList{
private:
    DoublyNode* head;
    DoublyNode* tail;
    int size;
public:

    DoublyLinkedList(){
        this->head = nullptr;
        this->tail = nullptr;
        this->size = 0;
    }

    int get_size(){
        return this->size;
    }

    DoublyNode* get_head(){
        return this->head;
    }
    DoublyNode* get_tail(){
        return this->tail;
    }

    void insert_front(int value){
        DoublyNode* newNode = new DoublyNode(value,this->head);

        this->head = newNode;
        this->size++;
        if(this->tail == nullptr){
            this->tail = this->head;
        }
    }

    void insert_last(int value){
        
        DoublyNode* newNode = new DoublyNode(value);
        if(this->head == nullptr){
            this->head = newNode;
            this->tail = newNode;
            this->size++;
            return;
        }

        this->tail->next = newNode;
        newNode->prev = this->tail;
        this->tail = newNode;
        this->size++;
    }

    int insert_middle(int value, int index){
        if(index>this->size){
            cout<<"index out of bound\n";
            return -1;
        }
        if(index==0){
            insert_front(value);
            return 0;
        }

        if(index==this->size){
            insert_last(value);
            return 0;
        }

        int pos = 0;
        DoublyNode* current = this->head;
        DoublyNode* latter = current->next;
        DoublyNode* newNode = new DoublyNode(value);

        while(pos<index-1){
            current = current->next;
            latter = current->next;
            pos++;
        }
        newNode->next = latter;
        latter->prev=newNode;

        current->next = newNode;
        newNode->prev = current;

        this->size++;
        return 0;
        
    }


    void remove_front(){
        if(this->head==nullptr && this->tail==nullptr) return;
        
        DoublyNode* temp = this->head;
        if(this->head == this->tail){
            this->tail == nullptr;
        }
        this->head = this->head->next;
        this->head->prev = nullptr;
        delete temp;
        this->size--;
    }

    void remove_last(){
        if(this->head==nullptr && this->tail==nullptr) return;
        DoublyNode* current = this->tail;

        if(this->head == this->tail){
            this->head == nullptr;
        }
        this->tail  = this->tail->prev;
        delete current;
        this->size--;
    }

    int remove_middle(int index){
        if(index>=this->size){
            cout<<"Index out of bound\n";
            return -1;
        }

        if(index == 0){
            this->remove_front();
            return 0;
        }

        if(index==this->size-1){
            this->remove_last();
            return 0;
        }

        DoublyNode* current = this->head;
        int pos = 0;

        while(pos<index-1){
            current = current->next;
            pos++;
        }

        DoublyNode* temp = current->next;

        current->next = temp->next;
        temp->next->prev = current;

        delete temp;
        this->size--;
        return 0;
    }

    int find(int value){
        DoublyNode* current = this->head;
        int pos = 0;
        while(current!=nullptr && current->value!=value){
            current = current->next;
            pos++;
        }

        if(current!=nullptr){
            return pos;
        }
        else{
            return -1;
        }
    }

    int remove_with_value(int value){
        int index = this->find(value);

        if(index==-1){
            cout<<"Value not available to remove\n";
            return -1;
        }

        this->remove_middle(index);
        return 0;
    }

    void printForward(){
        DoublyNode* current = this->head;

        while(current!=NULL){
            cout<<current->value<<" <-> ";
            current = current->next;
        }cout<<'\n';
    }

    void printBackward(){
        DoublyNode* current = this->tail;

        while(current!=NULL){
            cout<<current->value<<" <-> ";
            current = current->prev;
        }cout<<'\n';
    }


    
};

DoublyLinkedList make_unique(DoublyLinkedList arr){
    DoublyLinkedList ans;
    DoublyNode* current = arr.get_head();
    DoublyNode* curr_ans = ans.get_head();

    while(current){
        if(current->prev== nullptr && current->value!=current->next->value ){
            ans.insert_last(current->value);
        }
        else if(current->next == nullptr && current->value!=current->prev->value){
            ans.insert_last(current->value);
        }
        else if(current->prev != nullptr && current->next != nullptr && current->value!=current->prev->value && current->value!=current->next->value){
            ans.insert_last(current->value);
        }
        current = current->next;
    }
    return ans;
}


int main(int argc, char const *argv[])
{
        // associates standard input with input.txt 
    freopen("Cin.txt", "r", stdin);
    // associates standard output with output.txt
    // (this will create a new file called output.txt if none exists)
	freopen("Cout.txt", "w", stdout);

    int n;  cin>>n;
    DoublyLinkedList arr;

    for (int i = 0; i < n; i++)
    {
        int a;  cin>>a;
        arr.insert_last(a);
    }

    DoublyLinkedList unique = make_unique(arr);

    unique.printForward();
    
    return 0;
}

File: Java/DoublyLinkedList.java
class DoublyNode{
    int value;
    DoublyNode next;
    DoublyNode prev;

    DoublyNode(int value){
        this.value = value;
        this.next=null;
        this.prev = null;
    }

    DoublyNode(int value, DoublyNode next){
        this.value=value;
        this.next = next;
        this.prev = null;
    }

    DoublyNode(int value, DoublyNode prev, DoublyNode next){
        this.value=value;
        this.prev=prev;
        this.next=next;
    }

    
}

public class DoublyLinkedList {
    private DoublyNode head;
    private DoublyNode tail;

    private int size;

    public int getSize(){
        return this.size;
    }

    public int getHead(){
        return this.head.value;
    }

    public int getTail(){
        return this.tail.value;
    }

    public DoublyLinkedList(){
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    public void insertAtBeginning(int value){
        DoublyNode newNode = new DoublyNode(value,this.head);
        if(this.head==null){
            this.tail = newNode;
        }
        else{
            this.head.prev = newNode;
        }
        this.head =newNode;
        this.size++;
    }

    public void insertAtEnd(int value){
        DoublyNode newNode = new DoublyNode(value,this.tail,null);
        if(this.tail==null){
            this.head = newNode;
        }
        else{
            this.tail.next = newNode;
        }
        this.tail = newNode;
        this.size++;
    }

    public void insertAtMiddle(int value, int index){
        if(index>this.size) return;

        if(index==0){
            insertAtBeginning(value);
            return;
        }

        if(index==this.size){
            insertAtEnd(value);
            return;
        }

        if(2*index<this.size){
            DoublyNode currNode = this.head;
            DoublyNode newNode = new DoublyNode(value);
            int count = 0;

            while(count!=index-1){
                count++;
                currNode = currNode.next;
            }
            newNode.prev = currNode;
            newNode.next = currNode.next;
            currNode.next.prev = newNode;
            currNode.next = newNode;
        }

        else{
            DoublyNode curNode = this.tail;
            DoublyNode newNode = new DoublyNode(value);
            int count = this.size-1;

            while(count!=index){
                count++;
                curNode = curNode.prev;
            }

            curNode.prev.next = newNode;
            newNode.prev = curNode.prev;
            curNode.prev = newNode;
            newNode.next = curNode;
        }

        this.size++;
    }

    public void removeAtBeginning(){
        if(this.head==null) return;
        this.head = this.head.next;
        this.head.prev = null;
        this.size--;
    }

    public void removeAtEnd(){
        if(this.tail==null) return;
        this.tail = this.tail.prev;
        this.tail.next = null;
        this.size--;
    }

    public void removeAtMiddle(int index){
        if(index>this.size) return;
        if(index==0){
            removeAtBeginning();
            return;
        }
        if(index==this.size-1){
            removeAtEnd();
            return;
        }

        if(2*index<this.size){
            DoublyNode currNode = this.head;
            int count =0;
            
            while(count!=index-1){
                count++;
                currNode = currNode.next;
            }

            currNode.next = currNode.next.next;
            currNode.next.prev = currNode;
        }

        else{
            DoublyNode currNode = this.tail;
            int count = this.size-1;

            while(count!=index+1){
                count--;
                currNode = currNode.prev;
            }

            currNode.prev = currNode.prev.prev;
            currNode.prev.next = currNode;
        }
        this.size--;
    }

    public int find(int value){
        DoublyNode currNode = this.head;
        int index=0;
        while(currNode!=null && currNode.value!=value){
            index++;
            currNode = currNode.next;
        }

        if(currNode==null)
            return -1;

        return index;
    }

    public int removeWithValue(int value){
        DoublyNode currNode = this.head;
        int index = 0;

        while(currNode!=null && currNode.value!=value){
            index++;
            currNode = currNode.next;
        }

        if(currNode==null) return -1;
        return index;
    }

    public void printForward(){
        DoublyNode currNode = this.head;
        while(currNode!=null){
            System.out.print(currNode.value+" <-> ");
            currNode = currNode.next;
        }System.out.println();
    }

    public void printBackward(){
        DoublyNode currNode = this.tail;

        while(currNode!=null){
            System.out.println(currNode.value+" <-> ");
            currNode = currNode.prev;
        }System.out.println();
    }
}
